drop firewalld monitoring, add better iptables monitor

danwinship opened the pull request:
**What this PR does / why we need it**:
If an administrator reloads the system firewall, it may result in k8s's iptables rules being deleted. We dealt with this in two different ways, but both were problematic:
1. When the system firewall was managed by firewalld, we listened for D-Bus signals to know when it had reloaded, so we could immediately reload our rules as well. But this code was not covered by e2e tests, and in the process of trying to add an e2e test for it, I realized that the current OpenShift 4.x CI runs in an environment (Red Hat CoreOS) without firewalld. So even if we added an e2e test for firewalld monitoring, it would get skipped by both upstream Kubernetes e2e and by OpenShift e2e. Meanwhile, the firewalld monitoring code is the only consumer of `pkg/util/dbus`, which in theory had unit tests but in practice also never actually got tested because no one was ever running the unit tests in an environment where they had access to D-Bus. (FWIW, the current firewalld-monitoring code *does* actually work though.)
2. To cover the non-firewalld case, kubelet and kube-proxy both periodically recreate all of their iptables rules if they didn't exist. (Once a minute for kubelet, every 30s by default for kube-proxy, but configurable.) The problem with this (as discussed some in #80368) is that running `iptables -t filter ...` requires downloading the entire `filter` table from the kernel regardless of what command you're doing, and likewise for `iptables -t nat`, so even just checking that a chain still exists can be quite slow on systems with lots of iptables rules and can end up interfering with more important iptables operations (eg, actually creating chains for new services).

So this PR drops the firewalld code, adds some better iptables monitoring code, and updates kubelet and kube-proxy to use that new code. The new code takes advantage of the fact that all major linux-distro firewall-managing systems will flush *all* iptables tables on reload, so rather than trying to monitor chains in `nat` and `filter`, we can detect when iptables has been flushed by monitoring one of the relatively-empty iptables tables. (This means we no longer recover from someone gratuitously deleting *just* kubernetes's iptables rules, but... Don't Do That Then!)

In theory `iptables.Monitor` could be made to automatically choose between watching firewalld or doing periodic polling, based on whether firewalld was running. But that would make the code more complicated, and, as above, no one would be testing it. Given that the existing non-firewalld-based iptables monitoring code was working without problems for most users before, and that the changes here should make it work better even in very large clusters, it didn't seem worth it to try to keep the firewalld code.

**Which issue(s) this PR fixes**:
Fixes #15636 (by completely getting rid of `pkg/util/dbus`)

**Does this PR introduce a user-facing change?**:
```release-note
Kubernetes no longer monitors firewalld. On systems using firewalld for firewall
maintenance, kube-proxy will take slightly longer to recover from disruptive
firewalld operations that delete kube-proxy's iptables rules.

As a side effect of these changes, kube-proxy's
`sync_proxy_rules_last_timestamp_seconds` metric no longer behaves the
way it used to; now it will only change when services or endpoints actually
change, rather than reliably updating every 60 seconds (or whatever). If you
are trying to monitor for whether iptables updates are failing, the
`sync_proxy_rules_iptables_restore_failures_total` metric may be more useful.
```

/kind cleanup
/sig network
/area kube-proxy
/priority important-soon

danwinship committed:
Drop iptables firewalld monitoring support

The firewalld monitoring code was not well tested (and not easily
testable), would never be triggered on most platforms, and was only
being taken advantage of from one place (kube-proxy), which didn't
need it anyway since it already has its own resync loop.

Since the firewalld monitoring was the only consumer of pkg/util/dbus,
we can also now delete that.

danwinship committed:
Add iptables.Monitor, use it from kubelet and kube-proxy

Kubelet and kube-proxy both had loops to ensure that their iptables
rules didn't get deleted, by repeatedly recreating them. But on
systems with lots of iptables rules (ie, thousands of services), this
can be very slow (and thus might end up holding the iptables lock for
several seconds, blocking other operations, etc).

The specific threat that they need to worry about is
firewall-management commands that flush *all* dynamic iptables rules.
So add a new iptables.Monitor() function that handles this by creating
iptables-flush canaries and only triggering a full rule reload after
noticing that someone has deleted those chains.

danwinship committed:
e2e: test that both kube-proxy and kubelet recover after iptables flush

danwinship commented:
/uncc dcbw dchen1107 kubernetes/dep-approvers

danwinship commented:
/retest

squeed commented:
It's also easy enough to flush a single table (e.g. `iptables -F` will only flush filter). Will this code detect that case?

It probably makes sense to add a canary chain to *every* table. Even more awesome would be just to detect that in the proxy's save-load cycle - too bad that's a pretty ugly abstraction violation.

danwinship commented:
> It's also easy enough to flush a single table (e.g. `iptables -F` will only flush filter).

It's also easy to `killall kube-proxy` or `ifdown eth0`. The goal here isn't to try to recover from arbitrary breakage, it's just to interoperate with standard distro firewall-management tools, which all assume that they are allowed to delete any iptables rules that they didn't create themselves.

> It probably makes sense to add a canary chain to _every_ table.

Note that in the current code, only the `mangle` canary actually gets polled. The others are only there to avoid a race condition (where we run the `reloadFunc` before the firewall tool has finished deleting things).

> Even more awesome would be just to detect that in the proxy's save-load cycle - too bad that's a pretty ugly abstraction violation.

kube-proxy will already recreate all of the chains it needs any time it updates anything. I guess in theory it could also recreate kubelet's chains (maybe that's the abstraction violation you meant). It doesn't have the state to recreate `KUBE-HOSTPORT` though. (Note that the hostport code is currently handling reloads anyway, but that's another bug...)

danwinship commented:
/assign @thockin

danwinship commented:
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/81517/pull-kubernetes-e2e-gce/1163252942589923336/build-log.txt shows a test run where the new test was not marked "Disruptive", and with additional debugging added. (Search for "iptables rules".) The iptables rule deletion doesn't entirely work because the other tests being run in parallel are adding new services while we're trying to delete all the service rules, but it still deletes enough to break things, and then recovers.

danwinship commented:
/cc @dcbw

thockin commented:
First, I want to thank you for an EXCELLENT PR comment.  If only every PR was a thorough at explaining itself.

It seems like the root of the problem is multiple agents who think they own iptables and can do what they want.  We have this.  We have that other open issue with Calico and kube-proxy wrestling over ACCEPT and REJECT.  I have heard reports of other things like LB systems wrestling with kube-proxy.

I'll go read this code now, but I wonder if we need to make a more fundamental contract between systems that need to engage iptables, and then teach e.g. firewalld to play nice with that?

danwinship commented:
> I wonder if we need to make a more fundamental contract between systems that need to engage iptables, and then teach e.g. firewalld to play nice with that?

Well, it's not just firewalld though. It's firewalld, and Fedora/RHEL's iptables.services, and Debian's 17 different firewall-managing systems, and whatever SUSE uses when it's not using firewalld, etc. Trying to get things that are "below" kubernetes to play along with us may be hard but we can try to do better with things that are "above" us. I'd suggested before that k8s could create empty chains at various well-known points in the packet-processing process which plugins, etc, could make use of: https://groups.google.com/d/msg/kubernetes-sig-network/Hw9mmMoF_88/lf9Mv0jeAwAJ

I also had the idea at one point that if kube-proxy ran in its own network namespace, that might make things a lot simpler. It would have a veth connecting it to the main network namespace, and would set up a route so that all traffic to the service CIDR would be directed through that veth. Then those packets would get run through iptables rules *in kube-proxy's network namespace*, and after they'd been rewritten, they'd get redirected back out the veth to the main network namespace again where they could be routed to their final destination. (This solves the iptables-nft vs iptables-legacy problem too.)

I'm not totally sure it would work though, and it doesn't help with nodeport/externalip/etc which would still need to be handled in the root namespace...

danwinship commented on pkg/util/iptables/iptables.go:
```diff
@@ -524,6 +481,73 @@ func (runner *runner) checkRuleUsingCheck(args []string) (bool, error) {
+			// Our canary in the "mangle" table has been deleted; wait for the
+			// other canaries to be deleted too before returning so we don't
+			// start reloading too soon.
+			err := utilwait.PollImmediate(iptablesFlushPollTime, iptablesFlushTimeout, func() (bool, error) {
```
How would the caller know what a good timeout and period are though? How can the caller, in isolation, know what a good "low-impact" table is, given that this depends on what rules *other people* are creating? What the caller wants is just "if iptables gets flushed, call me", and `utiliptables` should be responsible for figuring out the best way to implement that for them.

So of the parameters:
- polling interval: I made this an argument to `Monitor` mostly because if I didn't then kube-proxy's `iptablesSyncPeriod` would become a no-op. But still, I think unlike the other stuff, this *is* something the caller might care about. Different callers might have different tolerances for how long they're willing to be out of sync.
- trigger table (`"mangle"`) - This needs to be decided based on the system as a whole. We could autodetect a good table by scanning `iptables-save` output, or we could add kubelet and kube-proxy config arguments to let the admin configure it based on their knowledge of their system. But mostly, the problem we're trying to work around is that kubernetes itself makes the `nat` and `filter` tables get huge, and I haven't heard of any systems where that's also a problem with `mangle`, `raw`, or `security`. I picked `mangle` somewhat at random from the three. (`security` seemed like a bad idea.) If we always used `mangle` now and that turned out to be a problem we could do the autodetect thing later.
- canary chain name (`KUBE-CANARY-XXXXXX`) - We don't let people rename any of our other chains... I guess maybe you're arguing that `pkg/util/iptables` is too low-level to be imposing the `KUBE-` prefix on the caller, and maybe that's true...
- post-trigger-table-flush poll interval (`100ms`) and timeout (`5s`) - I totally pulled these numbers out of my ass. I'm happy to revisit them but I don't think we need to make them configurable. All callers want the same behavior here ("respond quickly; don't get stuck if something goes wrong").

  FTR, basically all of the firewall managers do the same thing when reloading:

      for table in $(cat /proc/net/ip_tables_names); do
          iptables -F -t $table
          iptables -X -t $table
      done

  So we're just trying to wait for that loop to complete, which normally should happen very quickly, but maybe sometimes might take a bit longer, but shouldn't ever take *very* long.
- other tables to wait for (`["nat", "filter"]`) - OK, you got me, I am totally cheating there. But we could fix this without the caller having to pass any more information; the `runner` already knows what tables the user cares about based on the `EnsureChain`/`EnsureRule`/`Restore` commands they have run, so we can just figure out what tables to wait for based on that.

thockin replied:
> trigger table (`"mangle"`) - This needs to be decided based on the system as a whole. 
> We could autodetect a good table by scanning `iptables-save` output, or we could add 
> kubelet and kube-proxy config arguments to let the admin configure it based on their 
> knowledge of their system. But mostly, the problem we're trying to work around is that 
> kubernetes itself makes the `nat` and `filter` tables get huge, and I haven't heard of any 
> systems where that's also a problem with `mangle`, `raw`, or `security`. I picked `mangle` 
> somewhat at random from the three. (`security` seemed like a bad idea.) If we always used 
> `mangle` now and that turned out to be a problem we could do the autodetect thing later.

My thinking here was that the knowledge that "kubernetes itself makes the `nat` and `filter` tables get huge" doesn't belong in a nearly-general-purpose lib like this.  I could easily see moving this lib to a new repo k/iptables or just k/utils/iptables.  At least kube-proxy and kubelet know that kubernetes (ab)uses nat and filter, so they are in a position to give a hint.  Maybe the hint is "use mangle" or "use anything except nat or filter" instead? 

Auto-detecting sounds good, but I worry a little about point-in-time samples (e.g. at machine boot time, nat might be a good choice, because kube-proxy hasn't spun up yet.  So I'd still say we need a hint, at least.  "not nat, not filter."

> canary chain name (`KUBE-CANARY-XXXXXX`) - We don't let people rename any of our 
> other chains... I guess maybe you're arguing that `pkg/util/iptables` is too low-level to be 
> imposing the `KUBE-` prefix on the caller, and maybe that's true...

Yes, I don't think this should have "KUBE" coded in, and I assume the rand is because you can have multiple agents Monitor()ing, but I thin rand is kind of busted (as I said elsewhere).

The rest I don't care so much :)

> other tables to wait for (`["nat", "filter"]`) - OK, you got me, I am totally cheating there. But we 
> could fix this without the caller having to pass any more information; the `runner` already 
> knows what tables the user cares about based on the `EnsureChain`/`EnsureRule`/`Restore` > commands they have run, so we can just figure out what tables to wait for based on that.

I would expect Monitor() to be called very early, so kube-proxy might not have called any Ensure yet.  Parsing Restore blobs seems clunky to me.  It could be that the "blockers list" *is* the canary-table hint.  Don't canary in blocker tables.

danwinship commented on pkg/util/iptables/iptables.go:
```diff
@@ -524,6 +481,73 @@ func (runner *runner) checkRuleUsingCheck(args []string) (bool, error) {
+// Monitor is part of Interface
+func (runner *runner) Monitor(reloadFunc func(), interval time.Duration, stopCh <-chan struct{}) {
+	bytes := make([]byte, 5)
+	rand.Read(bytes)
```
Yes, it will leak chains, but I'm not sure why you think it will force a reload... ?

The leak is unfixable the way things are now, since we can't tell the difference between a chain that was leaked by a previous run of the same agent vs a chain that is still in use by another process. So I guess that's an argument for caller-passes-in-chain-name!

thockin commented:
> I'd suggested before that k8s could create empty chains at various well-known points in the packet-processing process which plugins, etc, could make use of

I like this idea more than ever, but *JUST* kubernetes will not convince others to use it.

Do we need a "LSB-for-iptables" where we define rules like "Each intrinsic chain FOO will jump to FOO-X, FOO-Y, FOO-Z in that order.  All DENY/REJECT rules should go in X, all ACCEPT rules should go in Z."  And then undertake a wider linux-community campaign to get people to respect that.

This is hard...

danwinship commented:
> It could be that the "blockers list" is the canary-table hint. Don't canary in blocker tables.

That happens to work for the case of kubelet and kube-proxy, but it doesn't make sense in general. Eg, kubelet only adds 3 rules to `filter` and 4 to `nat`, so in isolation it doesn't have any reason to think that those are going to be bad tables to poll.

I changed it to specify that it will poll the first table you pass as the list of tables to monitor. So now kubelet and kube-proxy both pass `["mangle", "filter", "nat"]` and it does what we want. Though I still think using `mangle` always would also make sense. Googling for examples of using the `mangle` table, you see lots of examples of mangling *all* packets, or all packets on a particular interface, all packets destined for a particular port, or all packets coming from a particular subnet, but none of those are likely to add up to tens of thousands of rules the way kube-proxy's `filter` and `nat` rules do.

Also, it turns out I was misremembering how bad "`iptables -t nat -C ...`" vs "`iptables -t mangle -C ...`" was. The latter *is* faster, but it's, eg, 0.004s vs 0.021s. The existing code is slow because kube-proxy does a full `iptables-restore` every sync period, not because it does an `iptables -C` on a bad choice table. So we *could* just get rid of this optimization and have `Monitor` only monitor the tables we *actualy* care about?

changes from before:
- Caller now passes canary chain name and list of tables to monitor to `Monitor`. The docs specifically call out the fact that the first table in the provided list will be the one that gets polled, so you can optimize by putting an empty table first. kubelet and kube-proxy both pass `["mangle", "filter", "nat"]`
- The polling code now uses `iptables -L` rather than `iptables -F` because the latter doesn't work as expected with iptables-nft in iptables <= 1.8.2.
- It now makes even less effort to delete the canary chains on error, because with the non-randomized names, there's less concern with "leaking"
- Added comments and logging
- Moved kubelet's `ipt.Monitor()` call from `kubelet.go` to `kubelet_network_linux.go` because it needs to include the list of tables now and `kubelet.go` has no reason to know that.

/hold
(I added back the debugging commit to test the new code)

danwinship commented:
/retest
(Flake due to us having removed `[Disruptive]` from the disruptive reload test for testing purposes.)

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -65,10 +64,16 @@ type Interface interface {
+	// the polling by listing a relatively empty table in tables[0]). When a flush is
+	// detected, this calls the reloadFunc so the caller can reload their own iptables
+	// rules.
+	// (This function should be called from a goroutine.)
```
Document error handling (it early-returns on error, etc)

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -529,12 +491,79 @@ func (runner *runner) checkRuleUsingCheck(args []string) (bool, error) {
+			return true, nil
+		}, stopCh)
+
+		for _, table := range tables {
```
Do we need to delete it here?  Best case they are already gone as per the loop above.  Worst case they are not gone but we timed out and are proceeding anyway.  We'll just recreate them at the top of this loop.

This seems like a waste

thockin reviewed (commented):
Otherwise LGTM

danwinship commented:
Updated doc comment, fixed another comment, and moved the cleanup so it only happens in the `stopCh` case.

And I removed the debug commit, so
/hold cancel

thockin reviewed (commented):
Thanks!

/lgtm
/approve

tedyu commented on pkg/util/iptables/iptables.go:
```diff
@@ -529,12 +492,78 @@ func (runner *runner) checkRuleUsingCheck(args []string) (bool, error) {
+
+			// Wait for the other canaries to be deleted too before returning
+			// so we don't start reloading too soon.
+			err := utilwait.PollImmediate(iptablesFlushPollTime, iptablesFlushTimeout, func() (bool, error) {
```
It seems if len(tables) == 1, this call can be skipped.

squeed commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -323,7 +323,13 @@ func NewProxier(ipt utiliptables.Interface,
+	// We pass syncPeriod to ipt.Monitor, which will call us only if it needs to.
+	// We need to pass *some* maxInterval to NewBoundedFrequencyRunner anyway though.
+	// time.Hour is arbitrary.
+	proxier.syncRunner = async.NewBoundedFrequencyRunner("sync-runner", proxier.syncProxyRules, minSyncPeriod, time.Hour, burstSyncs)
```
I suspect this will break iptables healthz on quiescent clusters. It marks kube-proxy unheahlthy if iptables hasn't been successfully synced in more than 2*SyncPeriod seconds.
